src/core/mac/mac.cpp

```diff
@@ -694,24 +694,39 @@ void Mac::ProcessTransmitSecurity(Frame &aFrame)
     {
     case Frame::kKeyIdMode0:
         key = mNetif.GetKeyManager().GetKek();
-        frameCounter = mNetif.GetKeyManager().GetKekFrameCounter();
-        mNetif.GetKeyManager().IncrementKekFrameCounter();
         extAddress = &mExtAddress;
+
+        if (!aFrame.IsARetransmission())
+        {
+            aFrame.SetFrameCounter(mNetif.GetKeyManager().GetKekFrameCounter());
+            mNetif.GetKeyManager().IncrementKekFrameCounter();
+        }
+
         break;
 
     case Frame::kKeyIdMode1:
         key = mNetif.GetKeyManager().GetCurrentMacKey();
-        frameCounter = mNetif.GetKeyManager().GetMacFrameCounter();
-        mNetif.GetKeyManager().IncrementMacFrameCounter();
-        aFrame.SetKeyId((mNetif.GetKeyManager().GetCurrentKeySequence() & 0x7f) + 1);
         extAddress = &mExtAddress;
+
+        // If the frame is marked as a retransmission, the `Mac::Sender` which prepared the frame should set the frame
+        // counter and key id to the same values used in the earlier transmit attempt. For a new frame (not a
+        // retransmission), we get a new frame counter and key id from the key manager.
+
+        if (!aFrame.IsARetransmission())
+        {
+            aFrame.SetFrameCounter(mNetif.GetKeyManager().GetMacFrameCounter());
+            mNetif.GetKeyManager().IncrementMacFrameCounter();
+            aFrame.SetKeyId((mNetif.GetKeyManager().GetCurrentKeySequence() & 0x7f) + 1);
```

Comment:
I agree that the spec requires the retx to use same DSN.. I will go ahead and add it...
It will still work ok without it though (as long as we use the same frame counter...)..